apps/api/src/services/sendgrid_handler.py

```diff
@@ -0,0 +1,64 @@
+# using SendGrid's Python Library
+# https://github.com/sendgrid/sendgrid-python
+import os
+from logging import getLogger
+from typing import Iterable, Mapping, Tuple, Union
+
+import aiosendgrid
+from httpx import HTTPStatusError
+from sendgrid.helpers.mail import Email, Mail, Personalization
+
+log = getLogger(__name__)
+
+SENDGRID_API_KEY = os.getenv("SENDGRID_API_KEY")
+
+PersonalizationData = Mapping[str, object]
+
+
+async def send_email(
+    template_id: str,
+    sender_email: Tuple[str, str],
+    receiver_data: Union[PersonalizationData, Iterable[PersonalizationData]],
+    send_to_multiple: bool = False,
+) -> None:
+    """
+    Send a personalized templated email to one or multiple receivers via SendGrid
+    """
+    try:
+        email_message = Mail()
+
+        if send_to_multiple:
+            if not isinstance(receiver_data, list):
+                raise TypeError(
+                    f"Expected {list} for receiver_data but got {type(receiver_data)}"
+                )
+            else:
+                for r in receiver_data:
```

Comment:
Issue: since we declared the parameter as type `Union[PersonalizationData, Iterable[PersonalizationData]]` but then assert instance of `list`, mypy coerces the type to `list[Any]` and the types of arguments to the `Email` constructor aren't checked.
While keeping `Iterable` is the most flexible approach, we may need to switch to `list` or use a typing overload to specify two versions based on `send_to_multiple`.

Reply:
Would we somehow have to adjust the stub so that it detects `send_to_multiple`?

Reply:
Interestingly Pylance is smart enough and correctly narrows the generic to `list[PersonalizationData]`.
Honestly, I don't think we actually use the `send_to_multiple` option, so maybe we could get rid of it?

Reply:
Don't we use `send_to_multiple` to check whether it's one email or multiple?

Reply:
In practice (in `email_handler`), we send only single emails, never multiple, and that was fine for last year. Not sure if we'll need this functionality this year though.

Reply:
We might need it for sending out acceptances

Reply:
Kept the option while also overloading in 703c09e. Still not perfect in narrowing but will work for now.